``Low Index Subgroups`` algorithm implementation in SymPy

gxyd opened the pull request:
Implements the "Low Index Subgroups" algorithm.

Commit:
``Low Index Subgroups`` algorithm implementation in SymPy

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -259,6 +284,53 @@ def scan(self, alpha, word):
+        if i > j:
+            if f != b:
+                return False
+            return True
```
Maybe `return f == b`.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -407,6 +479,33 @@ def process_deductions(self, R_c_x, R_c_x_inv):
+                alpha, x = self.deduction_stack.pop()
+                if p[alpha] == alpha:
+                    for w in R_c_x:
+                        self.scan_check(alpha, w)
```
We should probably have something like `if not self.scan_check(alpha, w): return False`.
It also seems that the next line will not be needed since `scan_check` does not modify `p`.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -407,6 +479,33 @@ def process_deductions(self, R_c_x, R_c_x_inv):
+                    self.scan_check(beta, w)
+                    if p[beta] < beta:
+                        break
+
```
If the loop ends successfully with empty stack, we should return `True`.

Commit:
code fixes in LowIndexSubgroup algorithm

Commit:
is_complete routine's docs improved

Commit:
improvements in LowIndexSubgroup algorithm

Commit:
Add docs for first_int_class

Commit:
Add doctest for first_int_class, copy() routine corrected

Commit:
Use a flag instead of putting a function in first_int_class

Set a flag, break from the inner loop and test it at an outer level

Commit:
name correction first_int_class -> first_in_class, change lamda

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -125,16 +135,34 @@ def omega(self):
-        return not None in flatten(self.table)
+        """
+        The coset table is called complete if it has no undefined entries
+        on the live cosets; that is, whether α^x is defined for all
```
'whether' could be omitted.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -259,6 +286,61 @@ def scan(self, alpha, word):
+            # we have an incorrect completed scan with coincidence f ~ b
+            # return False, instead of calling coincidence routine
+            return False
+        else:
```
This is the place where we can profitably extend the table by processing a deduction if `i == j`. That will then effectively complete the scan by 'joining' the forwards and backwards scans.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -688,4 +798,410 @@ def coset_enumeration_c(fp_grp, Y):
+
+    """
+    n = C.n
+    lamda = -1
```
This probably could be 0. Then we could have `range(lambda)`. And `lambda += 1` should be moved a bit later.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -688,4 +798,410 @@ def coset_enumeration_c(fp_grp, Y):
+    n = C.n
+    lamda = -1
+    nu = [None]*n
+    mu = []
```
This could probably also be `[None]*n`. Then there would be no need to `try` and `append`.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -688,4 +798,410 @@ def coset_enumeration_c(fp_grp, Y):
+        # compare corresponding entries in C and C_α
+        lamda = 0
+        for beta in range(C.n):
+            if breaker:
```
Perhaps this could be after the following loop. And probably `breaker` should be reset. (Something like `next_alpha` might be more descriptive from a mathematicians point of view.)

Commit:
Address @jksuom's comment and add comments

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -259,6 +285,65 @@ def scan(self, alpha, word):
+            # return True otherwise
+            self.table[f][A_dict[word[i]]] = b
+            self.table[b][A_dict_inv[word[i]]] = f
+        else:
```
Here we should probably have only `return True` (also in case `i == j``).

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -688,4 +798,445 @@ def coset_enumeration_c(fp_grp, Y):
+                    next_alpha = True
+                    break
+            if next_alpha:
+                break
```
It seems that `next_alpha` should be reset before `break` (`next_alpha = False`).

gxyd replied:
Yes, definitely.

On 06/17/2016 03:57 PM, Kalevi Suominen wrote:
>
> In sympy/combinatorics/fp_groups.py 
> <https://github.com/sympy/sympy/pull/11231#discussion_r67490420>:
>
> > +                    # continue with α
> > +                    next_alpha = True
> > +                    break
> > +                if nu[delta] is None:
> > +                    # delta becomes the next point in Ω_C_α
> > +                    lamda += 1
> > +                    nu[delta] = lamda
> > +                    mu[lamda] = delta
> > +                if nu[delta] < gamma:
> > +                    return False
> > +                if nu[delta] > gamma:
> > +                    # continue with α
> > +                    next_alpha = True
> > +                    break
> > +            if next_alpha:
> > +                break
>
> It seems that |next_alpha| should be reset before |break| (|next_alpha 
> = False|).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sympy/sympy/pull/11231/files/09134253eb3492d3304a8f0ebd223dbecddcdda1#r67490420>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AHWt8Snol98x9yBiE-VxN7bMn3-YSwPqks5qMnaugaJpZM4IzwKC>.
>

Commit:
set next_alpha to False, when next_alpha turns True

gxyd commented:
I have now basic understanding of `first_in_class`, added the doctest for confirmation of correct working of `first_in_class` routine, probably now it is working fine (apart from `lamda` comment, though that is not an issue result wise). Example has been taken from Derek Holt's handbook. Will now see for other routines, few comments to be addressed regarding `scan_check` and `process_deductions_check` routines.

jksuom commented:
`first_in_class` looks good.

Commit:
Furnish the code in routines for LowIndexSubgroups algorithm

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -688,4 +807,374 @@ def coset_enumeration_c(fp_grp, Y):
+
+    bool: True/False
+
+    If this returns False, then no descendant of D can have that property, and
```
The coset table in this paragraph should probably be the argument `C`, not `D`.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -378,9 +464,9 @@ def scan_and_fill_f(self, alpha, word):
-            # without making new definitions
+        # complete scan all relators under all cosets(obviously live)
+        # without making new definitions
+        for beta, w in self.omega:
```
Should `w` really be here?

gxyd replied:
No that shouldn't be here. Its a wrong change. I was trying to use 
`product` here, and realized that there is to be a use of `continue with β`.

On 06/18/2016 04:06 PM, Kalevi Suominen wrote:
>
> In sympy/combinatorics/fp_groups.py 
> <https://github.com/sympy/sympy/pull/11231#discussion_r67599158>:
>
> > @@ -378,9 +464,9 @@ def scan_and_fill_f(self, alpha, word):
> >      def look_ahead(self):
> >          R = self.fp_group.relators()
> >          p = self.p
> > -        for beta in self.omega:
> > -            # complete scan all relators under all cosets(obviously live)
> > -            # without making new definitions
> > +        # complete scan all relators under all cosets(obviously live)
> > +        # without making new definitions
> > +        for beta, w in self.omega:
>
> Should |w| really be here?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sympy/sympy/pull/11231/files/cf9ce8dfab3b29154db79c917b8e733148045e2c#r67599158>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AHWt8UbGvW6Uzbh3uwF_xPRe3RI0epZSks5qM8oagaJpZM4IzwKC>.
>

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -688,4 +807,374 @@ def coset_enumeration_c(fp_grp, Y):
+        for beta in reach:
+            if beta == C.n:
+                C.table.append([None]*len(C.A))
+                C.p.append(len(C.A) - 1)
```
Should this be `len(C.p)`?

gxyd replied:
Yes, this should be.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -688,4 +807,374 @@ def coset_enumeration_c(fp_grp, Y):
+    D.table[beta][D.A_dict_inv[x]] = alpha
+    D.deduction_stack.append((alpha, x))
+    D.process_deductions_check(R1_c_list[D.A_dict[x]], R1_c_list[D.A_dict_inv[x]])
+    if D.n == 0:
```
Alternatively, we could have `process_deductions_check` return a boolean. Then there should be no need to modify the property `n`.

gxyd replied:
Yes, that will be good from the implementation point of view. (though not from mathematician point of view, I believe). I agree this should be changed. :)

Commit:
undo the try of product use in "lookahead" routine

Commit:
Fix the wrong code in "descendant_subgroups" routine

Add tests for "low_index_subgroup" algorithm and also
fix the use of len(C.A)-1 -> len(C.p) in descendant_subgroups

gxyd commented:
Seems like Python2 doesn't have attribute `copy` for a `list` instance. Will fix that.

Commit:
use "list()" for copying a list, fix "C" not "D"

gxyd commented:
Also regarding the previous comment on removing the deductions from the `scan_check` routine. I saw in the solved example the use of "making deductions", and apart from `scan_check` there can be no other routine which can "make deductions", so thought of restoring it, and it seems to be working fine.

gxyd commented:
Will it be useful to have a routine `is_valid` for `CosetTable`? which will check for whether the instance of `CosetTable` is indeed possible considering its attribute value of `p`, `table`, since `.table`, `.p` can be changed. For example `[[0, 0, 0, 0], [1, 1, 1, 1]]` is an in-valid `CosetTable` since it doesn't obey the property you described yesterday.

jksuom commented:
The actual `table` is probably not valid all the time since not all rows are in omega. Otherwise I believe the tables created by the algorithms are valid. Do you think there would be need for `is_valid` somewhere?

gxyd commented:
No, not as in such. But while a was testing the procedures, I found the 
need to use it while implementing the "coset enumeration" in PR #11140. 
<https://github.com/sympy/sympy/pull/11140>

But I don't think it would be too much necessary. Since that wouldn't be 
from the user point of view.

On 06/19/2016 02:55 PM, Kalevi Suominen wrote:
>
> The actual |table| is probably not valid all the time since not all 
> rows are in omega. Otherwise I believe the tables created by the 
> algorithms are valid. Do you think there would be need for |is_valid| 
> somewhere?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sympy/sympy/pull/11231#issuecomment-226987747>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe/AHWt8VG4IVDjnDkSb0wyan-3ofrBkyItks5qNQshgaJpZM4IzwKC>.
>

Commit:
Add the variations to the LowIndexSubgroup algorithm

gxyd commented:
I have added the variations, which are mentioned in the section 5.4.3 of Handbook.

Commit:
Improve doctests for `first_in_class` routine

Commit:
Add test for variation of low_index_subgroup algorithm

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -688,4 +783,222 @@ def coset_enumeration_c(fp_grp, Y):
+        # G_α corresponding to this preceding table definitely contains the
+        # given subgroup
+        for w in Y:
+            if C.table[alpha][C.A_dict[w]] != alpha:
```
It seems that we have to assume that the subgroup generators in `Y` (if there are some) are also among the generators of the whole group (or their inverses). Otherwise it is necessary to run something like `scan`.

Commit:
Add comment for later fixing the bug

gxyd commented:
I have added the comment.

jksuom commented:
I think this is ready to be merged.

gxyd commented:
Yes, I think so too.

On 06/21/2016 08:17 PM, Kalevi Suominen wrote:
>
> I think this is ready to be merged.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sympy/sympy/pull/11231#issuecomment-227463517>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe/AHWt8aqd7UR3EeOIj3N6-RpMibafzl1Dks5qN_megaJpZM4IzwKC>.
>
